File: morsehgp3D_v9/audits/c_alternatives_20260923/experiences/D6_gpu/tiles.cpp
// Statistiques de tuiles (D3/D6) : points par tuile cubique de cote T et par tuile elargie d'un halo h <= T.
// Sert a dimensionner une execution GPU par tuile en memoire partagee. Entree u32le (mm).
#include <cstdio>
#include <cstdint>
#include <vector>
#include <map>
#include <array>
#include <algorithm>
#include <tuple>
int main(int argc, char** argv) {
  for (int f = 1; f < argc; ++f) {
    FILE* fp = std::fopen(argv[f], "rb");
    std::vector<std::array<int64_t,3>> P;
    uint32_t t[3];
    while (std::fread(t, 4, 3, fp) == 3) P.push_back({t[0], t[1], t[2]});
    std::fclose(fp);
    const size_t n = P.size();
    std::printf("%s n=%zu\n", argv[f], n);
    const int64_t cfg[][2] = {{1000,1000},{2000,2000}};
    for (auto& c : cfg) {
      const int64_t T = c[0], h = c[1];
      std::map<std::tuple<int64_t,int64_t,int64_t>, std::vector<size_t>> tiles;
      for (size_t i = 0; i < n; ++i) tiles[{P[i][0]/T, P[i][1]/T, P[i][2]/T}].push_back(i);
      std::vector<size_t> core, tot;
      for (auto& [key, ids] : tiles) {
        auto [tx,ty,tz] = key;
        const int64_t lo[3] = {tx*T - h, ty*T - h, tz*T - h}, hi[3] = {(tx+1)*T + h, (ty+1)*T + h, (tz+1)*T + h};
        size_t s = 0;
        const int R = (int)((h + T - 1) / T); for (int dx=-R; dx<=R; ++dx) for (int dy=-R; dy<=R; ++dy) for (int dz=-R; dz<=R; ++dz) {
          auto it = tiles.find({tx+dx, ty+dy, tz+dz});
          if (it == tiles.end()) continue;
          for (size_t i : it->second) {
            bool in = true;
            for (int a = 0; a < 3; ++a) in = in && P[i][a] >= lo[a] && P[i][a] < hi[a];
            s += in;
          }
        }
        core.push_back(ids.size()); tot.push_back(s);
      }
      auto q = [](std::vector<size_t> v, double fr) { std::sort(v.begin(), v.end()); return v[std::min(v.size()-1, (size_t)(fr*v.size()))]; };
      double dup = 0; size_t big = 0, bigpts = 0, big8 = 0, big8pts = 0;
      for (size_t i = 0; i < tot.size(); ++i) {
        dup += tot[i];
        if (tot[i] > 2048) { ++big; bigpts += core[i]; }
        if (tot[i] > 8192) { ++big8; big8pts += core[i]; }
      }
      std::printf("  T=%lldmm h=%lldmm tuiles=%zu coeur p50/p90/p99/max=%zu/%zu/%zu/%zu  tuile+halo p50/p90/p99/max=%zu/%zu/%zu/%zu  "
                  "duplication=%.2f  tuiles>2048: %zu (%.1f%% des sites)  >8192: %zu (%.1f%% des sites)\n",
                  (long long)T, (long long)h, core.size(), q(core,.5), q(core,.9), q(core,.99), *std::max_element(core.begin(), core.end()),
                  q(tot,.5), q(tot,.9), q(tot,.99), *std::max_element(tot.begin(), tot.end()), dup / n,
                  big, 100.0*bigpts/n, big8, 100.0*big8pts/n);
    }
  }
}

File: morsehgp3D_v9/audits/c_alternatives_20260923/experiences/D6_gpu/tiles.py
# Statistiques de tuiles D3 pour dimensionner une execution GPU (memoire partagee 99 Kio/bloc).
# Entree : triplets u32 little-endian (mm). Aucune dependance externe.
import struct, sys, collections, math
def load(p):
    b=open(p,'rb').read(); n=len(b)//12
    return [struct.unpack_from('<3I',b,12*i) for i in range(n)]
def q(v,f):
    v=sorted(v); n=len(v)
    return v[min(n-1,int(f*n))]
for path in sys.argv[1:]:
    P=load(path); n=len(P)
    xs=[p[0] for p in P]; ys=[p[1] for p in P]; zs=[p[2] for p in P]
    print(path.split('/')[-1], 'n=',n, 'bbox_m=', [(max(a)-min(a))/1000 for a in (xs,ys,zs)])
    for T,h in ((250,200),(500,200),(500,500),(1000,500)):
        # tuiles cubiques de cote T mm ; halo h mm : on compte les points dans la tuile elargie (cube T+2h).
        tiles=collections.Counter((x//T,y//T,z//T) for x,y,z in P)
        # grille fine de pas g=50 mm pour sommer le halo exactement au pas pres (majorant)
        g=50
        fine=collections.Counter((x//g,y//g,z//g) for x,y,z in P)
        r=(h+g-1)//g
        tot=[];core=[]
        for (tx,ty,tz),c in tiles.items():
            s=0
            x0=(tx*T)//g-r; x1=((tx+1)*T-1)//g+r
            y0=(ty*T)//g-r; y1=((ty+1)*T-1)//g+r
            z0=(tz*T)//g-r; z1=((tz+1)*T-1)//g+r
            if (x1-x0+1)*(y1-y0+1)*(z1-z0+1) < len(fine):
                for X in range(x0,x1+1):
                    for Y in range(y0,y1+1):
                        for Z in range(z0,z1+1):
                            s+=fine.get((X,Y,Z),0)
            else:
                for (X,Y,Z),cc in fine.items():
                    if x0<=X<=x1 and y0<=Y<=y1 and z0<=Z<=z1: s+=cc
            tot.append(s); core.append(c)
        dup=sum(tot)/n
        print(f'  T={T}mm h={h}mm tuiles={len(tiles)} coeur p50/p99/max={q(core,.5)}/{q(core,.99)}/{max(core)} '
              f'tuile+halo p50/p99/max={q(tot,.5)}/{q(tot,.99)}/{max(tot)} duplication={dup:.2f} '
              f'part_tuiles_halo>2048={sum(1 for t in tot if t>2048)/len(tot):.3f} points_dans_ces_tuiles={sum(c for c,t in zip(core,tot) if t>2048)/n:.3f}')
